README.md
# MiniCheck

MiniCheck provides a simple Rack application for adding simple health checks to your app.
The JSON output is similar to the one provided by the [Metrics](http://metrics.codahale.com/) Java library.
It was started at [Workshare ltd.](http://www.workshare.com) as an easy way of providing monitoring to our Rack based applications.

[![Gem Version](https://badge.fury.io/rb/mini_check.png)](http://badge.fury.io/rb/mini_check)
[![Build Status](https://secure.travis-ci.org/workshare/mini-check.png)](http://travis-ci.org/workshare/mini-check)

## Quick Start

Install the gem with the usual `gem install mini_check`.
Build a new Rack app and register some checks

```ruby
MyHealthCheck = MiniCheck::RackApp.new(path: '/healthcheck')
MyHealthCheck.register('health.redis_client'){ MyApp.redis_client.connected? }
MyHealthCheck.register('health.db_connection'){ MyApp.db_connection.fetch('show tables').to_a }
```

Mount it in your `config.ru`:

```ruby
use MyHealthCheck
run MyApp
```

If you now visit `http://localhost:XXXX/healthcheck` you should get something like:

```json
{
  "health.db_connection": {
    "healthy": true,
    "time": 0.111403303
  },
  "health.redis_client": {
    "healthy": true,
    "time": 4.41e-05
  }
}
```

The registered lambdas should do any of the following things:

* Return true if the check was successful.
* Return false if not.
* Raise an exception which will be understood as not healthy. Find an example of the output below:

```json
{
  "health.db_connection": {
    "healthy": false,
    "error": {
      "message": "Mysql2::Error: MySQL server has gone away",
      "stack": [
        "/home/manuel/sd/my_app/vendor/bundle/ruby/1.9.1/gems/sequel-4.7.0/lib/sequel/adapters/mysql2.rb:77:in `query'",
        "/home/manuel/sd/my_app/vendor/bundle/ruby/1.9.1/gems/sequel-4.7.0/lib/sequel/adapters/mysql2.rb:77:in `block in _execute'",
        "/home/manuel/sd/my_app/vendor/bundle/ruby/1.9.1/gems/sequel-4.7.0/lib/sequel/database/logging.rb:37:in `log_yield'",
        "..."
      ],
      "time": 4.61e-05
    }
  },
  "health.redis_client": {
    "healthy": true,
    "time": 1.23e-04
  }
}
```

The http status code will be 200 if all checks are healthy and 500 otherwise.

## Version

```ruby
MyVersionCheck = MiniCheck::VersionRackApp.new(name: 'Cards', path: '/admin/version', build_file: './config/build.yml').tap do |app|
    app.metadata["Whatever Here"] = "Bla Bla"
    ...
end
```

The build_file can be a YML or a plain text file. It needs to have pairs key-values.

If you now visit `http://localhost:XXXX/admin/version.json` you should get something like:

```json
{
  "Application Name": "Cards",
  "Whatever Here": "Bla Bla"
}
```

If you now visit `http://localhost:XXXX/admin/version` you should get something like:

```
  Application Name=Cards
  Whatever Here=Bla Bla
```

## Maintaining

Here is a good tutorial on this:
[Developing a RubyGem using Bundler](https://github.com/radar/guides/blob/master/gem-development.md).

## License

Released under the MIT License.  See the [LICENSE](LICENSE.md) file for further details.



CHANGELOG.md
## 0.4.1, release 2019-01-27

* Fixed broken build due Rake upgrade.
* Removed Code Climate as current implementation is being deprecated.
* Removed warning messages when running the tests.
* No code changes were made.

## 0.4.0

* Times on checks and slight refactor

## 0.3.0

* Extend gem to support mounting a new endpoint to provide versioning info.

## 0.2.0

* Added support for middleware mounting.

## 0.1.1, release 2014-03-22

* Supporting Travis CI and Code Climate

## 0.1.0, release 2014-03-09

* First working version.



spec/helper.rb
require 'rubygems'
require 'bundler'

begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end

require 'rspec'
require 'pry'

RSpec.configure do |config|
   config.tty = true
   config.formatter = :documentation
end

$LOAD_PATH.unshift File.expand_path('lib')
require 'mini_check'

$LOAD_PATH.unshift File.expand_path('spec/support')


spec/mini_check_spec.rb
require_relative 'helper'
require 'mini_check'

describe MiniCheck do
end


spec/mini_check/check_spec.rb
require_relative '../helper'

shared_examples_for 'check' do
  it{ is_expected.to respond_to(:to_hash) }
  it{ is_expected.to respond_to(:name) }
  it{ is_expected.to respond_to(:healthy?) }
  it{ is_expected.to respond_to(:run) }
end

describe MiniCheck::Check do
  subject{ MiniCheck::Check.new name: name, action: action }
  let(:name){ 'my_check' }
  let(:action){ proc{ true } }
  let(:exception){ Exception.new('My message') }

  it_behaves_like 'check'

  describe 'initialize' do
    it 'allows passing attributes as a hash' do
      check = MiniCheck::Check.new name: name, action: action
      expect(check.name).to eq(name)
      expect(check.action).to eq(action)
    end

    it 'allows building by name and proc' do
      check = MiniCheck::Check.new(name, &action)

      expect(check.name).to eq(name)
      expect(check.action).to eq(action)
    end
  end

  describe 'run' do
    it 'calls the action' do
      expect(action).to receive(:call)
      subject.run
    end

    context 'when the action returns an object (success)' do
      before :each do
        allow(action).to receive(:call).and_return(Object.new)
      end

      it 'sets the healthy? to true' do
        subject.healthy = nil
        subject.run
        expect(subject.healthy?).to eq(true)
      end

      it 'blanks the exception' do
        subject.exception = Exception.new
        subject.run
        expect(subject.exception).to be_nil
      end
    end

    context 'when the action returns false' do
      before :each do
        allow(action).to receive(:call).and_return(false)
      end

      it 'sets the healthy? to false' do
        subject.healthy = true
        subject.run
        expect(subject.healthy?).to eq(false)
      end

      it 'blanks the exception' do
        subject.exception = Exception.new
        subject.run
        expect(subject.exception).to be_nil
      end
    end

    context 'when the action raises an exception' do
      before :each do
        allow(action).to receive(:call).and_raise(exception)
      end

      it 'sets the healthy? to false' do
        subject.healthy = true
        subject.run
        expect(subject.healthy?).to eq(false)
      end

      it 'sets the exception to the exception' do
        subject.exception = nil
        subject.run
        expect(subject.exception).to be(exception)
      end
    end
  end

  describe 'to_hash' do
    before do
      allow_any_instance_of(Benchmark::Tms).to receive(:real).and_return(1.0)
    end

    context 'when the run was successful' do
      before :each do
        allow(action).to receive(:call).and_return(Object.new)
      end

      it 'returns the basic healthy hash' do
        subject.run
        expect(subject.to_hash).to eq(healthy: true, time: 1.0)
      end
    end

    context 'when the action returns false' do
      before :each do
        allow(action).to receive(:call).and_return(false)
      end

      it 'returns the basic healthy hash' do
        subject.run
        expect(subject.to_hash).to eq(healthy: false, time: 1.0)
      end
    end

    context 'when the action raises an exception' do
      before :each do
        allow(exception).to receive(:backtrace).and_return ['a']
        allow(action).to receive(:call).and_raise(exception)
      end

      it 'returns the hash with error' do
        subject.run
        expect(subject.to_hash).to eq(
          {
            healthy: false,
            error: {
              message: exception.message,
              stack: exception.backtrace
            },
            time: 1.0
          }
        )
      end
    end
  end
end


lib/mini_check/check.rb
require 'benchmark'

module MiniCheck
  class Check
    attr_accessor :name
    attr_accessor :healthy
    attr_accessor :action
    attr_accessor :time
    attr_accessor :exception

    def initialize args = {}, &block
      args = {name: args} if !args.is_a?(Hash)
      args[:action] = block if block_given?

      set_attributes args
    end

    def healthy?
      !!healthy
    end

    def run
      self.time = Benchmark.measure do
        begin
          do_run
          self.exception = nil
        rescue Exception => e
          self.healthy = false
          self.exception = e
        end
      end.real
    end

    def to_hash
      {}.tap do |h|
        h[:healthy] = healthy?
        h[:time] = time
        h[:error] = error_hash if exception
      end
    end

    private

    def do_run
      self.healthy = action.call
    end

    def error_hash
      {
        message: exception.message,
        stack: exception.backtrace
      }
    end

    def set_attributes args = {}
      args.each do |k,v|
        send("#{k}=", v)
      end
    end
  end
end


lib/mini_check/rack_app.rb
module MiniCheck
  class RackApp
    attr_accessor :checks
    attr_accessor :path

    CONTENT_TYPE_HEADER = 'Content-Type'.freeze
    JSON_MIME_TYPE = 'application/json'.freeze
    REQUEST_METHOD = 'REQUEST_METHOD'.freeze
    PATH_INFO = 'PATH_INFO'.freeze

    def initialize args = {}
      set_attributes args
    end

    def checks
      @checks ||= ChecksCollection.new
    end

    def call env
      case "#{env[REQUEST_METHOD]} #{env[PATH_INFO]}"
      when "GET #{path}"
        checks.run
        [status, default_headers, [body]]
      else
        host_app.call(env)
      end
    end

    def register name, &block
      checks.register name, &block
    end

    def new(app)
      dup.tap do |copy|
        copy.host_app = app
      end
    end

    private

    def default_headers
      { CONTENT_TYPE_HEADER => JSON_MIME_TYPE }
    end

    def set_attributes args = {}
      args.each do |k,v|
        send("#{k}=", v)
      end
    end

    def body
      checks.to_hash.to_json
    end

    def status
      checks.healthy? ? 200 : 500
    end

    protected

    attr_accessor :host_app

    def host_app
      @host_app ||= lambda { |env|  [404, {}, []] }
    end
  end
end


lib/mini_check/version_rack_app.rb
require 'yaml'

module MiniCheck
  class VersionRackApp
    attr_accessor :path, :build_file

    REQUEST_METHOD = 'REQUEST_METHOD'.freeze
    PATH_INFO = 'PATH_INFO'.freeze
    APP_KEY = 'Application Name'.freeze
    CONTENT_TYPE_HEADER = 'Content-Type'.freeze
    JSON_MIME_TYPE = 'application/json'.freeze
    TEXT_MIME_TYPE = 'text/plain'.freeze

    def initialize args = {}
      set_attributes args
    end

    def metadata
      @metadata ||= Hash.new
    end

    def call env
      case "#{env[REQUEST_METHOD]} #{env[PATH_INFO]}"
      when "GET #{path}.json"
        JsonResponse.render(output_hash)
      when "GET #{path}"
        PlainTextResponse.render(output_hash)
      else
        host_app.call(env)
      end
    end

    def new(app)
      copy = self.dup
      copy.host_app = app
      copy
    end


    private

    def name= name
      metadata[APP_KEY] = name
    end

    def set_attributes args = {}
      args.each do |k,v|
        send("#{k}=", v)
      end
    end

    def output_hash
      metadata.merge(file_hash)
    end

    def file_hash
      if (content = file_content(build_file)).instance_of?(Hash)
        content
      else
        content = raw_file_content(build_file).split("\n")
        Hash[content.map{ |pair| pair.split("=") }]
      end
    rescue => ex
      { error: ex.message }
    end

    def self.file_content(build_file)
      @file_content ||= YAML.load_file(build_file)
    end

    def self.raw_file_content(build_file)
      @raw_file_content ||= IO.read(build_file)
    end

    def file_content(build_file)
      self.class.file_content(build_file)
    end

    def raw_file_content(build_file)
      self.class.raw_file_content(build_file)
    end

    protected

    attr_accessor :host_app

    def host_app
      @host_app ||= lambda { |env|  [404, {}, []] }
    end


    class JsonResponse
      class << self
        def render hash
          [200, default_headers, [hash.to_json]]
        end

        private

        def default_headers
          { CONTENT_TYPE_HEADER => JSON_MIME_TYPE }
        end
      end
    end

    class PlainTextResponse
      class << self
        def render hash
          [200, default_headers, [parse(hash)]]
        end

        private

        def parse hash
          hash.map{ |key,value| "#{key}=#{value}" }.join("\n")
        end

        def default_headers
          { CONTENT_TYPE_HEADER => TEXT_MIME_TYPE }
        end
      end
    end
  end
end